tests/testsuite/publish.rs

```diff
@@ -2048,3 +2049,187 @@ error: package ID specification `bar` did not match any packages
         )
         .run();
 }
+
+#[cargo_test]
+fn wait_for_publish() {
+    // Counter for number of tries before the package is "published"
+    let arc: Arc<Mutex<u32>> = Arc::new(Mutex::new(0));
+    let arc2 = arc.clone();
+
+    // Registry returns an invalid response.
+    let _registry = registry::RegistryBuilder::new()
+        .http_index()
+        .http_api()
+        .add_responder("/index/de/la/delay", move |req, server| {
+            let mut lock = arc.lock().unwrap();
+            *lock += 1;
+            // if the package name contains _ or
+            if *lock <= 1 {
+                server.not_found(req)
+            } else {
+                server.index(req)
+            }
+        })
+        .build();
+
+    // The sparse-registry test server does not know how to publish on its own.
+    // So let us call publish for it.
+    Package::new("delay", "0.0.1")
+        .file("src/lib.rs", "")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "delay"
+                version = "0.0.1"
+                authors = []
+                license = "MIT"
+                description = "foo"
+
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("publish --no-verify -Z sparse-registry")
+        .masquerade_as_nightly_cargo(&["sparse-registry"])
+        .with_status(0)
+        .with_stderr(
+            "\
+[UPDATING] `dummy-registry` index
+[WARNING] using `registry.token` config value with source replacement is deprecated
+This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/xxx>.
+Use the --token command-line flag to remove this warning.
+[WARNING] manifest has no documentation, [..]
+See [..]
+[PACKAGING] delay v0.0.1 ([CWD])
+[UPLOADING] delay v0.0.1 ([CWD])
+[UPDATING] `dummy-registry` index
+[WAITING] on registry `dummy-registry` to finish publish
+[UPDATING] `dummy-registry` index
+",
+        )
+        .run();
+
+    // Verify the responder has been pinged
+    let lock = arc2.lock().unwrap();
+    assert_eq!(*lock, 2);
+    drop(lock);
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                [dependencies]
+                delay = "0.0.1"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("build -Z sparse-registry")
+        .masquerade_as_nightly_cargo(&["sparse-registry"])
+        .with_status(0)
+        .run();
+}
+
+/// A separate test is needed for package names with - or _ as they hit
```

Comment:
This is a bit unfortunate. Would it make sense to plumb the `QueryKind::Exact` all the way to `load_summaries()` and skip the `UncanonicalizedIter` for that case?

Reply:
imo this is an existing problem that we should address decoupled from this discussion.  I'd recommend opening an issue for us to track this.